[Workers] Add docs for Vitest testing (#13092)
cloudflare/cloudflare-docs: content/workers/testing/vitest/test-apis.md

---
title: Test APIs
pcx_content_type: reference
weight: 5
---

# Test APIs

> The Workers Vitest pool provides runtime helpers for writing tests in the `cloudflare:test` module. This module can only be imported from within test files run by the Workers Vitest pool.

Reviewer: ```suggestion
The Workers Vitest pool provides runtime helpers for writing tests in the `cloudflare:test` module. This module is provided by the `@cloudflare/vitest-pool-workers` package, but can only be imported from test files that execute in the Workers Vitest pool.
```

Can't decide if clever or odd how we've done this, but regardless — should be clear about what is providing it.

## `cloudflare:test` module definition

{{<definitions>}}

- {{<code>}}env: {{<type>}}import("cloudflare:test").ProvidedEnv{{</type>}}{{</code>}}

>   - 2nd argument passed to modules-format exported handlers. Contains bindings configured in top-level `miniflare` pool options. To configure the type of this value, use an ambient module type:

Reviewer: ```suggestion
  - Exposes the [`env` object](/workers/runtime-apis/handlers/fetch/#parameters) that provides access to [bindings](/workers/runtime-apis/bindings/) that you have [configured in `vitest.config.js`](/workers/testing/vitest/configuration/).
 
To configure the TypeScript type of `env`, use an ambient module type:
```

    <br>

>     ```ts

Reviewer: The first code example someone sees seems like it should be showing something like:

```js
import { env } from "cloudflare:test"
// ...show you how to write a test
```

    ---
    filename: env.d.ts
    ---
    declare module "cloudflare:test" {
      interface ProvidedEnv {
        NAMESPACE: KVNamespace;
      }
      // ...or if you have an existing `Env` type...
      interface ProvidedEnv extends Env {}
    }
    ```

- {{<code>}}SELF: {{<type>}}Fetcher{{</type>}}{{</code>}}

>   - [Service binding](/workers/runtime-apis/service-bindings/) to the default export defined in the `main` Worker. Use this to write integration tests against your Worker. Note this `main` Worker runs in the same isolate/context as tests so any global mocks will apply to it too.

Reviewer: In practice this might be a service binding under the hood, but I wouldn't expect someone to necessarily understand the implications?

It seems like the common use case for `SELF` is to use it to import the main Worker that's being tested, and call its [handlers](/workers/runtime-apis/handlers/) directly? Is that accurate? If so — can we say it?

- {{<code>}}fetchMock: {{<type>}}import("undici").MockAgent{{</type>}}{{</code>}}

>   - Declarative interface for mocking outbound `fetch()` requests. Deactivated by default and reset before running each test file. Refer to [`undici`'s `MockAgent` documentation](https://undici.nodejs.org/#/docs/api/MockAgent) for more information. Note this only mocks `fetch()` requests for the current test runner worker. Auxiliary workers should mock `fetch()`es using the Miniflare `fetchMock`/`outboundService` options.

Reviewer: "Auxiliarly workers"?  Define?

    <br>

    ```ts
    ---
    filename: index.spec.js
    ---
    import { fetchMock } from "cloudflare:test";
    import { beforeAll, afterEach, it, expect } from "vitest";

    beforeAll(() => {
      // Enable outbound request mocking...
      fetchMock.activate();
      // ...and throw errors if an outbound request isn't mocked
      fetchMock.disableNetConnect();
    });
    // Ensure we matched every mock we defined
    afterEach(() => fetchMock.assertNoPendingInterceptors());

    it("mocks requests", async () => {
      // Mock the first request to `https://example.com`
      fetchMock
        .get("https://example.com")
        .intercept({ path: "/" })
        .reply(200, "body");

      const response = await fetch("https://example.com/");
      expect(await response.text()).toBe("body");
    });
    ```

{{</definitions>}}

### Events

{{<definitions>}}

- {{<code>}}createExecutionContext(): {{<type>}}ExecutionContext{{</type>}}{{</code>}}

>   - Creates an instance of `ExecutionContext` for use as the 3rd argument to modules-format exported handlers.

Reviewer: ```suggestion
  - Creates an instance of the [`context` object](/workers/runtime-apis/handlers/fetch/#parameters)
```

+ show example of this API in use?

- {{<code>}}waitOnExecutionContext(ctx:{{<param-type>}}ExecutionContext{{</param-type>}}): {{<type>}}Promise\<void>{{</type>}}{{</code>}}

>   - Waits for all `ExecutionContext#waitUntil()`ed `Promise`s to settle. Only accepts instances of `ExecutionContext` returned by `createExecutionContext()`.

Reviewer: Explain why and how to use this — ex: "Use this to wait for all promises passed to `context.waitUntil()` to settle, before running test assertions."

  - Gets the acknowledged/retry state of messages in the `MessageBatch`, and waits for all `ExecutionContext#waitUntil()`ed `Promise`s to settle. Only accepts instances of `MessageBatch` returned by `createMessageBatch()`, and instances of `ExecutionContext` returned by `createExecutionContext()`.

    <br>

    ```ts
    ---
    filename: index.spec.js
    ---
    import { env, createMessageBatch, createExecutionContext, getQueueResult } from "cloudflare:test";
    import { it, expect } from "vitest";
    import worker from "./index.mjs";

    it("calls queue handler", async () => {
      const batch = createMessageBatch("my-queue", [
        {
          id: "message-1",
          timestamp: new Date(1000),
          body: "body-1"
        }
      ]);
      const ctx = createExecutionContext();
      await worker.queue(batch, env, ctx);
      const result = await getQueueResult(batch, ctx);
      expect(result.retryAll).toBe(false);
      expect(result.ackAll).toBe(false);
      expect(result.explicitRetries).toStrictEqual([]);
      expect(result.explicitAcks).toStrictEqual(["message-1"]);
    });
    ```

{{</definitions>}}

> ### Durable Objects

Reviewer: @vy-ton re: testing content for DO docs, or any cross-linking.

@kentonv @geelen — wondering how this will interact with the RPC API.

{{<definitions>}}

- {{<code>}}runInDurableObject\<O extends DurableObject, R>(stub:{{<param-type>}}DurableObjectStub{{</param-type>}}, callback:{{<param-type>}}(instance: O, state: DurableObjectState) => R | Promise\<R>{{</param-type>}}): {{<type>}}Promise\<R>{{</type>}}{{</code>}}

>   - Runs `callback` inside the Durable Object pointed-to by `stub`'s context. Conceptually, this temporarily replaces your Durable Object's `fetch()` handler with `callback`, then sends a request to it, returning the result. This can be used to call/spy-on Durable Object instance methods or seed/get persisted data. Note this can only be used with `stub`s pointing to Durable Objects defined in the `main` worker.

Reviewer: ```suggestion
  - Runs the provided `callback` inside the Durable Object instance that corresponds to the provided `stub`. 

This temporarily replaces your Durable Object's `fetch()` handler with the `callback` you provide and then sends a request to it, returning the result. This can be used to call/spy-on Durable Object instance methods or seed/get persisted data. Note this can only be used with stubs pointing to Durable Objects defined in the `main` worker.
```
